### DISCUSSION TITLE
Multiple databases in a multi-tenant architecture

### DISCUSSION THREAD

**Discussion by AJK74** [cloudnative-pg/cloudnative-pg#2357] [Q&A]
I note from the [FAQ](https://cloudnative-pg.io/documentation/1.16/faq/#database-management) that the recommended architecture is a single PostgreSQL database per cluster.

My application is SaaS with a relatively small number of customer instances (10s currently). 

Due to various constraints (that I can't change in the short-term) I'm stuck with each customer instance needing a relatively small (>4Gb) and fairly lightly used (1-2 active users at any time) PostgreSQL database - I'm planning for around 100 databases, all with the same data-schema. 

Much of the rest of the application is hosted in Kubernetes, and the bits that are not we'd like to move that way. 

I'm thinking in these circumstances I treat these 100 identical databases as if they were just one - and pool them on a single cluster so we can more easily manage resilience backups upgrades etc.  

Can you see any particular pitfalls with this approach that I'd need to be aware of and any particular limits you'd advise?

Assuming this is not a big no-no: 
It would be helpful for adopters in a position such as this if it was possible via the operator to optionally define databases (with roles) separately from the creation of the cluster - we can then more easily manage them alongside the other provisioning processes when a new customer instance is defined. 

Thanks!

**Comment by sxd** [comment]
Hi @AJK74 

Can you be a bit more clear on what you one to do? you want to have a 100 databases in one instance or merge all those 100 database into one? is not clear your final goal.

On the PostgreSQL, it's possible but not recommended, since if one database goes down, all of them will go down, instead, having one cluster with one database, it's harder to have a big outage when something goes wrong, that's why we strongly recommend to have one database per cluster and not all of them in just one cluster.

I'm also turning this into a discussion since it's not an issue.

Regards!

**Comment by AJK74** [comment]
Thanks sxd,

The goal for now is to be able to run 100 databases (or more).

I can't merge them together due to external dependencies that I can't change at the moment (the databases are used by a third party product).
 
All have the same data-schema, all are fairly small, and all are relatively lightly loaded with low user counts.

I'm thinking my use case is a poor fit for the way your operator is architected.  Which is a bit of a shame as many of your other features do align well (clustering, barman dr)

So instead I'm favoring setting up some larger clusters (one to start with but can add more as needed). 

I've started to test a bitnami-ha helm deployment which will require a lot more manual effort to configure up front, but I can shape this to our use case more easily.

Unless i'm missing something?!

Many thanks

**Comment by georgebarbarosie** [comment]
I will add to the voice against multi-tenant clusters with our experience.

We went through this transition ourselves, from all databases in one cluster to the current state where we spawn a cluster for each install. The transition to the current state was enabled by postgresql operators (we've just started transitioning production workloads to cloudnative-pg, from a different operator). We've operated in both paradigms.

With the cluster-per-workload approach, you can:
- decouple workloads' PITR needs:
  * recovery per workload, you don't need to replay all your largest workloads for one tiny database that needs to go back in time
  * backup/retention policies and frequency, choice of storage, all can be configured per workload
- easily scale horizontally
- allocate resources based on workload needs
- port your workloads from cluster to cluster easier

We would never contemplate going back to monolithic clusters today. Treat your database clusters as local to the workload, it's well worth it.

**Comment by simonmacklin** [comment]
I agree with the above statement, but I would be interested in the cloud cost comparisons between the two approaches.  I think for smaller, greenfield projects having one cluster per microservice makes perfect sense.  

But for larger enterprises with hundreds or even thousands of services that require a DB, having something in between might be a better approach to try and balance infra / cloud costs with best practices.

I have always used the one DB per cluster rule until recently, as you grow the costs way up even if you have lots of tiny clusters.  You will still likely need a standby and every cluster will have a baseline overhead and cost.

I think it would be very cool if the operator could manage multiple databases, roles ect to support all use cases
